File: leetcode/rust-leetcode/src/105.construct-binary-tree-from-preorder-and-inorder-traversal.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;
impl Solution {
    pub fn build_tree(preorder: Vec<i32>, inorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        if preorder.is_empty() {return None;}
        let mut map = HashMap::new();
        for (i, &n) in inorder.iter().enumerate() {
            map.insert(n, i);
        }
        fn build(preorder: &Vec<i32>, map: &HashMap<i32,usize>, start: usize, end: usize, idx: &mut usize) -> Option<Rc<RefCell<TreeNode>>> {
            if start > end {
                return None;
            }
            let val = preorder[*idx];
            let inorder_idx = *map.get(&val).unwrap();
            *idx += 1;
            let mut tnode = TreeNode::new(val);
            if start < inorder_idx {
                tnode.left = build(preorder, map, start, inorder_idx - 1, idx);
            }
            tnode.right = build(preorder, map, inorder_idx + 1, end, idx);
            Some(Rc::new(RefCell::new(tnode)))
        }
        let mut idx = 0;
        build(&preorder, &map, 0, inorder.len()-1, &mut idx)
    }
}

File: leetcode/go/105.construct-binary-tree-from-preorder-and-inorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
    if len(preorder) == 0 {
        return nil
    }
    m := make(map[int]int, len(inorder))
    for i, n := range inorder {
        m[n] = i
    }
    preOrderIdx := 0
    return build(preorder, &preOrderIdx, m, 0, len(preorder)-1)
}

func build(preorder []int, preOrderIdx *int, m map[int]int, start int, end int) *TreeNode {
    if start > end {
        return nil
    }
    rootVal := preorder[*preOrderIdx]
    root := TreeNode {
        Val: rootVal,
    }
    *preOrderIdx++
    if idx, ok := m[rootVal]; ok {
        root.Left = build(preorder, preOrderIdx, m, start, idx-1)
        root.Right = build(preorder, preOrderIdx, m, idx+1, end)
    }
    return &root
}

